src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp

```diff
@@ -514,22 +519,59 @@ MergeTreePrefetchedReadPool::ThreadsTasks MergeTreePrefetchedReadPool::createThr
 
             read_task->priority = priority;
 
-            bool allow_prefetch = !settings.filesystem_prefetches_limit || current_prefetches_count + 1 <= settings.filesystem_prefetches_limit;
-            if (allow_prefetch && settings.filesystem_prefetch_max_memory_usage)
+            bool allow_prefetch = false;
+            if (allowed_memory_usage
+                && (allowed_prefetches_num.has_value() == false || allowed_prefetches_num.value() > 0))
             {
-                size_t num_readers = 1;
+                /// adjustBufferSize(), which is done in MergeTreeReaderStream and MergeTreeReaderCompact,
+                /// lowers buffer size if file size (or required read range) is less. So we know that the
+                /// settings.prefetch_buffer_size will be lowered there, therefore we account it here as well.
+                /// But here we make a more approximate lowering, while in adjustBufferSize it will be presize.
+                size_t estimated_memory_usage = 0;
+                size_t estimated_prefetches_num = 0;
+
+                for (const auto & col : part.task_columns.columns)
```

Comment:
it isn't necessarily a problem but we're better to calculate this only once per part